src/Common/FileCache.cpp

```diff
@@ -0,0 +1,623 @@
+        file_segments = {cell->file_segment};
+    }
+    else
+    {
+        /// There are segments [segment1, ..., segmentN]
+        /// (non-overlapping, non-empty, ascending-ordered) which (maybe partially)
+        /// intersect with given range.
+
+        /// It can have holes:
+        /// [____________________]         -- requested range
+        ///     [____]  [_]   [_________]  -- intersecting cache [segment1, ..., segmentN]
+        ///
+        /// For each such hole create a cell with file segment state EMPTY.
+
+        auto it = file_segments.begin();
+        auto segment_range = (*it)->range();
+
+        size_t current_pos;
+        if (segment_range.left < range.left)
+        {
+            ///    [_______     -- requested range
+            /// [_______
+            /// ^
+            /// segment1
+
+            current_pos = segment_range.right + 1;
+            ++it;
+        }
+        else
+            current_pos = range.left;
+
+        while (current_pos <= range.right && it != file_segments.end())
+        {
+            segment_range = (*it)->range();
+
+            if (current_pos == segment_range.left)
+            {
+                current_pos = segment_range.right + 1;
+                ++it;
+                continue;
+            }
+
+            assert(current_pos < segment_range.left);
+
+            auto hole_size = segment_range.left - current_pos;
+            auto * cell = addCell(key, current_pos, hole_size, FileSegment::State::EMPTY, cache_lock);
+            file_segments.insert(it, cell->file_segment);
+
+            current_pos = segment_range.right + 1;
+            ++it;
+        }
+
+        if (current_pos <= range.right)
+        {
+            ///   ________]     -- requested range
+            ///   _____]
+            ///        ^
+            /// segmentN
+
+            auto hole_size = range.right - current_pos + 1;
+            auto * cell = addCell(key, current_pos, hole_size, FileSegment::State::EMPTY, cache_lock);
+            file_segments.push_back(cell->file_segment);
+        }
+    }
+
+    /// TODO: remove this extra debug logging.
+    String ranges;
+    for (const auto & s : file_segments)
+        ranges += "\nRange: " + s->range().toString() + ", download state: " + FileSegment::stateToString(s->download_state) + " ";
+    LOG_TEST(log, "Cache getOrSet. Key: {}, range: {}, file_segments number: {}, ranges: {}",
+             keyToStr(key), range.toString(), file_segments.size(), ranges);
+
+    assert(!file_segments.empty());
+    return FileSegmentsHolder(std::move(file_segments));
+}
+
+LRUFileCache::FileSegmentCell * LRUFileCache::addCell(
+    const Key & key, size_t offset, size_t size, FileSegment::State state,
+    std::lock_guard<std::mutex> & /* cache_lock */)
+{
+    /// Create a file segment cell and put it in `files` map by [key][offset].
+
+    if (!size)
+        return nullptr; /// Empty files are not cached.
+
+    if (files[key].contains(offset))
```

Comment:
As far as I understand `addCell` function is low-level and not for user. So we don't check for some containing or intersecting intervals. Maybe make sense to add such checks under `#ifndef NDEBUG`.

Reply:
> So we don't check for some containing or intersecting intervals.

No, it is checked, the checks are covered by a lot of asserts.